Add a redis library to generalize DB functions we need in SearXNG.

return42 opened the pull request:
## What does this PR do?

Add a redis library to generalize DB functions we need in SearXNG.

## Why is this change important?

- single point of definition
- we do not need multiple e.g. counter functions that stores key/values in the DB (this PR is inspired by https://github.com/searxng/searxng/pull/998 and https://github.com/searxng/searxng/pull/618)
- to not pollute the redis DB with keys, we preifx all keys from the SearXNG schema with `SearXNG_` (there is a searx.redislib.purge_by_prefix() function to drop the SearXNG data)

## How to test this PR locally?

In the second commit, the limiter plugin is modified to use `searx.redislib.incr_counter` .. the limiter is a good test case, further compile the documentation and have a look at `src/searx.redislib.html`.

## Author's checklist

I made several tests .. also with the limiter plugin .. @dalf I am not a lua expert but I think the lua script from the (old) limiter plugin is compareable / replaceable by the `searx.redislib.incr_counter`. May I have overseen something / could you place have a look at the second commit .. thanks!

return42 committed:
[mod] collection of redis functions and lua scripts (initial)

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] plugins/limiter.py - use searx.redislib.incr_sliding_window

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on searx/plugins/limiter.py:
```diff
@@ -47,70 +48,44 @@ def is_accepted_request(inc_get_counter) -> bool:
-
-def create_inc_get_counter(redis_client, secret_key_bytes):
-    lua_script = """
-    local slidingWindow = KEYS[1]
```
FYI: slidingWindow is an argument not a KEY.

return42 commented on searx/plugins/limiter.py:
```diff
@@ -47,70 +48,44 @@ def is_accepted_request(inc_get_counter) -> bool:
-    local key = KEYS[2]
-    local now = tonumber(redis.call('TIME')[1])
-    local id = redis.call('INCR', 'counter')
-    if (id > 2^46)
-    then
-        redis.call('SET', 'count', 0)
-    end
```
Instead of `id` I use the concatenation of `current_time[1] .. current_time[2]`

dalf commented on searx/redislib.py:
```diff
@@ -0,0 +1,201 @@
+      1
+
+    """
+    script = client.register_script(INCR_SLIDING_WINDOW)
```
The current pattern is:
* to call `redis_client.script_load` once when the application starts
* to call `return redis_client.evalsha(script_sha, ...)` afterward, so the script is not uploaded to Redis each time.

---

Here the function creates a `Script` instance for each call.
Why not defined the script as a global variable?

For reference the `Script` class (from https://redis.readthedocs.io/en/v4.3.1/_modules/redis/commands/core.html?highlight=register_script )

```python

class Script:
    """
    An executable Lua script object returned by ``register_script``
    """

    def __init__(self, registered_client, script):
        self.registered_client = registered_client
        self.script = script
        # Precalculate and store the SHA1 hex digest of the script.

        if isinstance(script, str):
            # We need the encoding from the client in order to generate an
            # accurate byte representation of the script
            encoder = self.get_encoder()
            script = encoder.encode(script)
        self.sha = hashlib.sha1(script).hexdigest()
```

return42 replied:
> Here the function creates a Script instance for each call.

Yes, the price is the calculation of the sha1 value on each request .. not optimal.

> Why not defined the script as a global variable?

AFAIK there is no function or a Script-constructor to create Script objects without a redis-client ... wait, now  I see there is one in the  core command:

https://redis.readthedocs.io/en/v4.3.1/commands.html?highlight=register_client#redis.commands.cluster.RedisClusterCommands.register_script

@dalf: 

I haven't tested this core `register_script` if you want I can give it a try.

But one question remains: when should this Script objects created .. while import `searx.redislib`?

dalf replied:
> when should this Script objects created

https://github.com/searxng/searxng/blob/38403ee515ba09fb675bbc8d544ea579436aeee6/searx/webapp.py#L1414-L1416

or

https://github.com/searxng/searxng/blob/38403ee515ba09fb675bbc8d544ea579436aeee6/searx/search/__init__.py#L27-L36

?

return42 replied:
OK, tomorrow I will give [register_script](https://redis.readthedocs.io/en/v4.3.1/commands.html?highlight=register_client#redis.commands.cluster.RedisClusterCommands.register_script) a try / Implement a init() function for the `searx/redislib.py`.

return42 replied:
@dalf rebased this branch and pushed a commit on top to cache `Script` objects.

dalf commented on searx/redislib.py:
```diff
@@ -0,0 +1,201 @@
+local current_time = redis.call('TIME')
+
+redis.call('ZREMRANGEBYSCORE', name, 0, current_time[1] - expire)
+redis.call('ZADD', name, current_time[1], current_time[1] .. current_time[2])
```
The purpose of the ID is to make sure to count every request even if there are at the exact same timestamp.

return42 replied:
See https://github.com/searxng/searxng/pull/1047#discussion_r884233029

dalf commented on searx/plugins/limiter.py:
```diff
@@ -47,70 +48,45 @@ def is_accepted_request(inc_get_counter) -> bool:
-        c_burst = inc_get_counter(interval=20, keys=[b'IP limit, burst', x_forwarded_for])
-        c_10min = inc_get_counter(interval=600, keys=[b'IP limit, 10 minutes', x_forwarded_for])
+        c_burst = incr_sliding_window(redis_client, 'IP limit, burst' + x_forwarded_for, 20)
+        c_10min = incr_sliding_window(redis_client, 'IP limit, 10 minutes' + x_forwarded_for, 600)
```
nitpicking: since `name` is not a Redis key anyway, it can be tuple so we avoid a string concat.

return42 replied:
Not sure I understand you: the `name` argument is a string that identifies the counter / it is the name of the counter .. why should the name of a counter a tuple? (or did I misunderstood you?)

dalf commented on searx/redislib.py:
```diff
@@ -0,0 +1,201 @@
+PURGE_BY_PREFIX = """
+local prefix = tostring(ARGV[1])
+for i, name in ipairs(redis.call('KEYS', prefix .. '*')) do
+    redis.call('EXPIRE', name, 0)
```
Why not `redis.call('DEL', name)` ?

return42 replied:
EXPIRE removes the object asynchronous while DEL is executed synchronous.

EDIT: synchronous DB manipulation blocks the python execution loop.

dalf replied:
I couldn't find information about `DEL` vs `EXPIRE 0`.

According to [this micro benchmark](https://gist.github.com/dalf/d43a06ad4e1c5c3913ff57b3278da22a#file-redis_expire_vs_del-py),  `EXPIRE` is more or less equivalent to `DEL` is the best case scenario, otherwise `DEL` is faster especially when there are a lot of keys. May be, I have this point. Could you document the choice of EXPIRE instead of DEL?

return42 replied:
Its not about what command is faster .. its about transaction frames in the DB itself .. if there is an open transaction frame in the DB the `DEL` has to wait until the frame is closed, then `DEL` can delete the key .. when you use `EXPIRE` the call of the `EXPIRE` turns back immediately and the python command loop can continue .. the deletion (expire) of the key in the DB is managed by the DB internally.

In line 5 of `searx/redislib.py` I gave a link to [Bullet-Proofing Lua Scripts in RedisPy](https://redis.com/blog/bullet-proofing-lua-scripts-in-redispy/) .. its a very interesting article, may you like to read it.

dalf commented on searx/redislib.py:
```diff
@@ -0,0 +1,201 @@
+      (5, 1)
+
+    """
+    script = client.register_script(INCR_COUNTER)
```
See the comment about `incr_sliding_window`.

return42 replied:
See my comment above: why should the `name` of a counter a tuple?

dalf commented on searx/plugins/limiter.py:
```diff
@@ -36,8 +36,9 @@
-def is_accepted_request(inc_get_counter) -> bool:
+def is_accepted_request() -> bool:
     # pylint: disable=too-many-return-statements
+    redis_client = redisdb.client()
```
The Redis client implements:
* a retry helper: https://redis.readthedocs.io/en/v4.3.1/retry.html
* a connection pool: https://pypi.org/project/redis/ (see "Connection Pools")

IMO, we can create one client and share it between the different threads.

So, for example, `incr_sliding_window(client, name: str, duration: int)` becomes `incr_sliding_window(name: str, duration: int)`.

return42 replied:
I know, but I think it is better not to bind a library to *one* DB connection.  IMO: The management of the DB connections is out of the scope of a library that implements of functions. .. may be one day we have multiple connections or a plugin want to use its own connection.

----

EDIT: ...

> we can create one client and share it between the different threads.

More or less we do this already in the `global _client` variable in `shared/redis.py`:

https://github.com/searxng/searxng/blob/e93385e6de8d77b1e8285547d642ca9f0d4021dc/searx/shared/redisdb.py#L30-L36

We can discuss if `shared/redis.py` is the right place to define the **one** DB connection of SearXNG, but I don't wanted to change connection management in this PR.  In this PR I thought it is better to focus on the function library ... and the function library should not limited to a single DB connection.

dalf replied:
* `redisdb.client()` connects to `localhost:6379`.
* the instantiation of a Redis client for each request has a performance impact according to [this micro benchmark](https://gist.github.com/dalf/d43a06ad4e1c5c3913ff57b3278da22a#file-reuse_redis_client-py)

return42 replied:
> * `redisdb.client()` connects to `localhost:6379`.

It is connected to whatever redis.url from settings point to: `redis.Redis.from_url(get_setting('redis.url')) `

> * the instantiation of a Redis client for each request has a performance impact

Not sure I understand, but why do you think that we *"the instantiation of a Redis client for each request"* .. if you look at the code sequence shown above, there is a global variable `_client` that is set only once (if it is None).

Or do I missed something?

dalf replied:
I'm sorry I've read `redis.Client()` instead of `redisdb.client()`

dalf reviewed (commented):
See the comments

return42 commented on searx/redislib.py:
```diff
@@ -0,0 +1,201 @@
+    :return: value of the incremented counter
+    :type return: int
+
+    The implementation of the redis counter is the lua script from string
+    :py:obj:`INCR_SLIDING_WINDOW`.  The lua script uses `sorted sets in Redis`_
+    to implement a sliding window for the redis key ``SearXNG_counter_<name>``
+    (ZADD_).  The current TIME_ is used to score the items in the sorted set and
+    the time window is moved by removing items with a score lower current time
+    minus *duration* time (ZREMRANGEBYSCORE_).
+
+    The EXPIRE_ time (the duration of the sliding window) is refreshed on each
+    call (incrementation) and if there is no call in this duration, the sorted
+    set expires from the redis DB.
+
+    The return value is the amount of items in the sorted set (ZCOUNT_), what
+    means the number of calls in the sliding window.
+
+    .. _Sorted sets in Redis:
+       https://redis.com/ebook/part-1-getting-started/chapter-1-getting-to-know-redis/1-2-what-redis-data-structures-look-like/1-2-5-sorted-sets-in-redis/
+    .. _TIME: https://redis.io/commands/time/
+    .. _ZADD: https://redis.io/commands/zadd/
+    .. _EXPIRE: https://redis.io/commands/expire/
+    .. _ZREMRANGEBYSCORE: https://redis.io/commands/zremrangebyscore/
+    .. _ZCOUNT: https://redis.io/commands/zcount/
```
> The purpose of the ID is to make sure to count every request even if there are at the exact same timestamp.

Not sure I understand your comment, but here in the doc-string the LUA script is described.  The [TIME](https://redis.io/commands/time/) returns tuple and the `..` is a concatenation of the two values in this tuple (to get an unique name).

return42 committed:
[mod] redislib - optimize LUA script registration.

Instead of register the LUA script on each call, the LUA script is registered
once and the `Script` object is cached in global storage.

[1] https://github.com/searxng/searxng/pull/1047#discussion_r884143305

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 commented on searx/redislib.py:
```diff
@@ -0,0 +1,241 @@
+    zero.  The default prefix will drop all keys which has been set by SearXNG
+    (drops SearXNG schema entirely from database).
+
+    The implementation is the lua script from string :py:obj:`PURGE_BY_PREFIX`.
+    The lua script uses EXPIRE_ instead of DEL_: if there are a lot keys to
+    delete and/or their values are big, `DEL` could take more time and blocks
+    the command loop while `EXPIRE` turns back immediate.
```
Here I added a description why we use `EXPIRE` instead of `DEL`.

dalf reviewed (approved):
Discussed on Matrix --> let's merge.
